File: unidbg-api/src/main/java/com/vortexdbg/spi/Loader.java
package com.vortexdbg.spi;

import com.vortexdbg.LibraryResolver;
import com.vortexdbg.Module;
import com.vortexdbg.ModuleListener;
import com.vortexdbg.Symbol;
import com.vortexdbg.hook.HookListener;
import com.vortexdbg.pointer.UnidbgPointer;

import java.io.File;
import java.util.Collection;
import java.util.Map;

@SuppressWarnings("unused")
public interface Loader {

    void setLibraryResolver(LibraryResolver libraryResolver);

    void disableCallInitFunction();

    void setCallInitFunction(boolean callInit);

    Module load(File elfFile);
    Module load(File elfFile, boolean forceCallInit);

    Module load(LibraryFile libraryFile);
    Module load(LibraryFile libraryFile, boolean forceCallInit);

    Module findModuleByAddress(long address);
    Module findModule(String name);

    Module dlopen(String filename);
    Module dlopen(String filename, boolean calInit);
    boolean dlclose(long handle);
    Symbol dlsym(long handle, String symbol);

    void addModuleListener(ModuleListener listener);

    void addHookListener(HookListener listener);

    Collection<Module> getLoadedModules();

    String getMaxLengthLibraryName();
    long getMaxSizeOfLibrary();

    /**
     * 加载虚拟模块
     */
    Module loadVirtualModule(String name, final Map<String, UnidbgPointer> symbols);

}

File: vortexdbg-api/src/main/kotlin/com/vortexdbg/spi/Loader.kt
package com.vortexdbg.spi

import com.vortexdbg.LibraryResolver
import com.vortexdbg.Module
import com.vortexdbg.ModuleListener
import com.vortexdbg.Symbol
import com.vortexdbg.hook.HookListener
import com.vortexdbg.pointer.VortexdbgPointer

import java.io.File

@Suppress("unused")
interface Loader {

    fun setLibraryResolver(libraryResolver: LibraryResolver)

    fun disableCallInitFunction()

    fun setCallInitFunction(callInit: Boolean)

    fun load(elfFile: File): Module
    fun load(elfFile: File, forceCallInit: Boolean): Module

    fun load(libraryFile: LibraryFile): Module
    fun load(libraryFile: LibraryFile, forceCallInit: Boolean): Module

    fun findModuleByAddress(address: Long): Module?
    fun findModule(name: String): Module?

    fun dlopen(filename: String): Module?
    fun dlopen(filename: String, calInit: Boolean): Module?
    fun dlclose(handle: Long): Boolean
    fun dlsym(handle: Long, symbol: String): Symbol?

    fun addModuleListener(listener: ModuleListener)

    fun addHookListener(listener: HookListener)

    fun getLoadedModules(): Collection<Module>

    fun getMaxLengthLibraryName(): String
    fun getMaxSizeOfLibrary(): Long

    /**
     * Registers a synthetic module backed by the given symbol table, without loading any file. Used
     * to expose host-provided implementations to guest code under a module name.
     */
    @JvmSuppressWildcards
    fun loadVirtualModule(name: String, symbols: Map<String, VortexdbgPointer>): Module

}
